src/webgpu/api/operation/vertex_state.spec.ts

```diff
@@ -0,0 +1,366 @@
+      }
+    }
+    return arrayBuffer;
+  }
+}
+
+export const g = makeTestGroup(IndexFormatTest);
+
+g.test('index_format,uint16')
+  .desc('Test rendering result of indexed draw with index format of uint16.')
+  .params([
+    { indexOffset: 0, _expectedShape: RenderShape.Square },
+    { indexOffset: 6, _expectedShape: RenderShape.BottomLeftTriangle },
+    { indexOffset: 18, _expectedShape: RenderShape.Nothing },
+  ])
+  .fn(t => {
+    const { indexOffset, _expectedShape } = t.params;
+
+    // If this is interpreted as uint32, it will have index 1 and 2 be both 0 and render nothing.
+    // And the index buffer size - offset must be not less than the size required by triangle
+    // list, otherwise it also render nothing.
+    const indexArray = new Uint16Array([1, 2, 0, 0, 0, 0, 0, 1, 3, 0]);
+    const indexBuffer = t.CreateIndexBuffer(indexArray, 'uint16');
+    const result = t.run(indexBuffer, indexArray.length, 'uint16', indexOffset);
+
+    const expectedTextureValues = t.CreateExpectedUint8Array(_expectedShape);
+    t.expectContents(result, expectedTextureValues);
+  });
+
+g.test('index_format,uint32')
+  .desc('Test rendering result of indexed draw with index format of uint32.')
+  .params([
+    { indexOffset: 0, _expectedShape: RenderShape.Square },
+    { indexOffset: 12, _expectedShape: RenderShape.BottomLeftTriangle },
+    { indexOffset: 36, _expectedShape: RenderShape.Nothing },
+  ])
+  .fn(t => {
+    const { indexOffset, _expectedShape } = t.params;
+
+    // If this is interpreted as uint16, then it would be 0, 1, 0, ... and would draw nothing.
+    // And the index buffer size - offset must be not less than the size required by triangle
+    // list, otherwise it also render nothing.
+    const indexArray = new Uint32Array([1, 2, 0, 0, 0, 0, 0, 1, 3, 0]);
+    const indexBuffer = t.CreateIndexBuffer(indexArray, 'uint32');
+    const result = t.run(indexBuffer, indexArray.length, 'uint32', indexOffset);
+
+    const expectedTextureValues = t.CreateExpectedUint8Array(_expectedShape);
+    t.expectContents(result, expectedTextureValues);
+  });
+
+g.test('primitive_restart')
+  .desc('Test primitive restart with each primitive topology.')
+  .params(
+    params()
+      .combine(poptions('indexFormat', ['uint16', 'uint32'] as const))
+      .combine(kPrimitiveTopologiesForRestart)
+  )
+  .fn(t => {
+    const { indexFormat, primitiveTopology, _expectedShape } = t.params;
+
+    let indexArray: Uint16Array | Uint32Array;
+    // The primitive restart value can used with strip primitive topologies ('line-strip' or 'triangle-strip').
+    // For lists, they are expected to be renderred the same as without primitive restart.
+    if (indexFormat === 'uint16') {
+      if (primitiveTopology === 'triangle-list') {
+        // triangle-list: (0, 1, 3), (0xffff, 2, 1)
+        // triangle-list with restart: (0, 1, 3), (2, 1, 0)
+        // triangle-strip: (0, 1, 3), (2, 1, 0)
```

Comment:
nit: Mark up which one is expected, like:
```suggestion
        // -> triangle-list: (0, 1, 3), (0xffff, 2, 1)
        //    triangle-list with restart: (0, 1, 3), (2, 1, 0)
        //    triangle-strip: (0, 1, 3), (2, 1, 0)
```
and same below.